features/cellular/framework/AT/ATHandler.cpp

```diff
@@ -120,8 +124,17 @@ bool ATHandler::get_debug() const
 
 ATHandler::~ATHandler()
 {
+    ScopedLock <ATHandler> lock(*this);
     set_file_handle(NULL);
 
+    while (_event_id > 0) {
+#ifdef AT_HANDLER_MUTEX
```

Comment:
I think we are in a bit of a pickle here.

We can't tell if cancel worked, so if we do call it we don't know if we need to wait for `event_id` to be cleared. That's why it got removed.

But if the OOBs are running on the same thread as this destructor, which I believe is quite possible in some event-driven apps, we *must* call cancel (and we can assume it would work in that case). We can't wait - we'd deadlock the event queue.

As a minimal change, this could be

    #ifdef AT_HANDLER_MUTEX
        while (_event_id > 0) {
            _oobCv.wait();
        }
    #else 
        if (_event_id > 0) { 
            event_cancel(_event_id);
        }
    #endif 

But that still isn't any good if using an event queue with AT_HANDLER_MUTEX defined.

Ideally I'd like a a success/error return from cancel. But I recall this was discussed before - @kivaisan was looking at it? I'm not sure if the conclusion was that it wasn't possible to get a solid return, or that a different solution was found.

Reply:
Is this something that would be better fixed by using the event's destructor?
https://github.com/ARMmbed/mbed-os/blob/master/events/equeue/equeue.h#L161

This tells you when the event has expired, either by completing or being canceled.

It's indirectly exposed through the C++ API via function objects, though it's messy:
```
bool dead = false;
class DeathHook {
    void operator()(void) {
        // do stuff
    }
    ~DeathHook() {
        dead = true;
    }
};

int id = queue.call(DeathHook());
queue.cancel(id);
printf("is dead? %d\n", dead);
```

Reply:
Does that example work at present? You're passing a temporary by value, and that temporary gets copied then destroyed, so I think dead is set immediately after the call().

I think it could work if EventQueue was upgraded to use C++11 perfect forwarding. Or, more simply, if the constructor/destructor reference counted.

Reply:
You know, I actually have no idea. I think you're right this would need reference counting to work.

Alternatively we could poke an explicit destructor-register function through the C++ API like in the C one.

Or you might be able to extend the Event class.